[incubator/elasticsearch] rbac, es v5, master and data resilience

icereval opened the pull request:
* rename files for consistency
* master as statefulset
* pre-stop-hook wait on shard evacuation
* RBAC support
* elasticsearch v5 w/ xpack plugin available [based](https://github.com/CenterForOpenScience/docker-library/blob/master/elasticsearch/5.4/Dockerfile) on the [official](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html) elastic image
* support elasticsearch 2

icereval committed:
rename files

icereval committed:
master as statefulset

icereval committed:
pre-stop-hook wait on shards

icereval committed:
RBAC support

icereval committed:
elasticsearch v5 w/ xpack plugin available

mikelorant commented on incubator/elasticsearch/values.yaml:
```diff
@@ -28,8 +32,10 @@ client:
-  replicas: 2
+  replicas: 3
   heapSize: "128m"
+  storage: "10Gi"
```
Recommend dropping this to 4G. Smallest EBS volume size and the metadata on the master is very little data.

mikelorant commented on incubator/elasticsearch/values.yaml:
```diff
@@ -50,7 +56,11 @@ data:
     limits:
       cpu: "1"
-      memory: "512Mi"
+      memory: "2048Mi"
```
Don't hardcode this limit. This is likely the mostly frequently tuned value.

icereval committed:
support elasticsearch 2

icereval committed:
changes from code review

icereval commented:
changes made from review, rebased on latest master branch (template namespacing changes applied)

totemcaf commented on incubator/elasticsearch/templates/master-statefulset.yaml:
```diff
@@ -9,6 +9,7 @@ metadata:
     release: {{ .Release.Name }}
   name: {{ template "elasticsearch.master.fullname" . }}
 spec:
+  serviceName: {{ template "master.fullname" . }}
```
"master.fullname" template si missing. Should it be "elasticsearch.master.fullname"

totemcaf commented on incubator/elasticsearch/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,19 @@
+kind: ClusterRoleBinding
+metadata:
+  labels:
+    app: {{ template "name" .}}
```
elasticsearch.name

totemcaf commented on incubator/elasticsearch/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,19 @@
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "fullname" . }}
```
elasticsearch.fullname

totemcaf commented on incubator/elasticsearch/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,19 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: {{ template "fullname" . }}
```
elasticsearch.fullname

totemcaf commented on incubator/elasticsearch/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,19 @@
+  name: {{ template "fullname" . }}
+subjects:
+- kind: ServiceAccount
+  name: {{ template "fullname" . }}
```
elasticsearch.fullname

totemcaf commented on incubator/elasticsearch/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,28 @@
+kind: ClusterRole
+metadata:
+  labels:
+    app: {{ template "name" . }}
```
elasticsearch.name

totemcaf commented on incubator/elasticsearch/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,28 @@
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ template "fullname" . }}
```
elasticsearch.fullname

totemcaf reviewed (changes requested):
I tried it locally and found some issues with template names.

icereval commented:
oops, @totemcaf thanks for the review, rebased on latest master as well.

yanns commented on incubator/elasticsearch/values.yaml:
```diff
@@ -3,45 +3,52 @@
 image:
-  repository: "jetstack/elasticsearch-pet"
-  tag: "2.4.0"
+  repository: "centerforopenscience/elasticsearch"
```
If you change the default container, please update the README. Thx!

icereval replied:
done

yanns replied:
👍

yanns replied:
Do we need a `LICENSE` for docker files?

yanns commented on incubator/elasticsearch/templates/configmap.yaml:
```diff
@@ -13,22 +13,42 @@ data:
       type: kubernetes
       zen:
-        minimum_master_nodes: 2
+        minimum_master_nodes: ${MINIMUM_MASTER_NODES:2}
```
👍

yanns commented on incubator/elasticsearch/values.yaml:
```diff
@@ -3,45 +3,52 @@
+  env:
+    # IMPORTANT: https://www.elastic.co/guide/en/elasticsearch/reference/current/important-settings.html#minimum_master_nodes
+    # To prevent data loss, it is vital to configure the discovery.zen.minimum_master_nodes setting so that each master-eligible
+    # node knows the minimum number of master-eligible nodes that must be visible in order to form a cluster.
```
Maybe we can add a FAQ like https://github.com/pires/kubernetes-elasticsearch-cluster#why-does-number_of_masters-differ-from-number-of-master-replicas

totemcaf commented:
@icereval I was finding info about the "client" role but I found a similar configuration with the name "Coordinating" role (https://www.elastic.co/guide/en/elasticsearch/reference/5.5/modules-node.html#coordinating-only-node). Can you add a reference to it? Tx

totemcaf commented:
@icereval Why is the port 9300 (Transport Protocol) is not enabled on Client Nodes?
Our microservices connect to the ES cluster using a Java Client that can benefit on using that port.

By the way, thanks a lot for your effort on upgrading the chart.

unguiculus commented on incubator/elasticsearch/templates/client-deployment.yaml:
```diff
@@ -63,6 +63,10 @@ spec:
               fieldPath: metadata.namespace
         - name: NODE_DATA
           value: "false"
+{{- if hasPrefix "5." .Values.image.tag }}
```
What if version 6 comes out or someone uses a custom image that has different versioning?

jmcarp replied:
@unguiculus: if version 6 comes out, we'll need to update the chart, but we'll probably need to update the chart for a major release anyway 😃. How would you prefer this to be configured? Would you rather the chart just support a single elastic version? Or do you think we should just special-case 2.x, like this:

```yaml
{{- if hasPrefix "2." .Values.image.tag }}
bootstrap.mlockall: ${BOOTSTRAP_MLOCKALL:false}
{{- else }}
bootstrap.memory_lock: ${BOOTSTRAP_MEMORY_LOCK:false}
{{- end }}
```

unguiculus replied:
Maybe you could add a separate value `esVersion` instead of relying on the image version. That would allow me to specify a custom image that is versioned differently.

icereval replied:
I went ahead and updated the chart with a values variable as suggested.

Looks like the concept of `appVersion` is already baked in to the `Chart.yaml`, unfortunately it is [immutable](https://github.com/kubernetes/helm/issues/2566#issuecomment-307964606).

more info:
- [The Chart.yaml File](https://github.com/kubernetes/helm/blob/1fa853959ccbdb1c2834bd625f3655b2a9e7a4cb/docs/charts.md#the-chartyaml-file)
- [The appVersion field](https://github.com/kubernetes/helm/blob/1fa853959ccbdb1c2834bd625f3655b2a9e7a4cb/docs/charts.md#the-appversion-field)

unguiculus commented on incubator/elasticsearch/templates/configmap.yaml:
```diff
@@ -13,22 +13,42 @@ data:
     
     node.data: ${NODE_DATA:true}
     node.master: ${NODE_MASTER:true}
+{{- if hasPrefix "5." .Values.image.tag }}
```
see above

unguiculus commented on incubator/elasticsearch/templates/configmap.yaml:
```diff
@@ -13,22 +13,42 @@ data:
         service: ${SERVICE}
         namespace: ${KUBERNETES_NAMESPACE}
 
+{{- if hasPrefix "2." .Values.image.tag }}
```
Again version specific. Add well-documented config options for these that do not rely on versions.

unguiculus commented on incubator/elasticsearch/templates/configmap.yaml:
```diff
@@ -13,22 +13,42 @@ data:
-        minimum_master_nodes: 2
+        minimum_master_nodes: ${MINIMUM_MASTER_NODES:2}
+{{- end }}
+{{- if hasPrefix "5." .Values.image.tag }}
```
see above

unguiculus commented on incubator/elasticsearch/templates/master-statefulset.yaml:
```diff
@@ -63,12 +77,20 @@ spec:
               fieldPath: metadata.namespace
         - name: NODE_DATA
           value: "false"
+{{- if hasPrefix "5." .Values.image.tag }}
```
see above

icereval commented:
@totemcaf - port 9300 added to client service

@totemcaf - it appears in v5 they removed reference the `client` node in documentation and now simply refer to it as `coordinating` node - I've added a note in the readme to clarify.

@yanns - I've added a note to the `Deep dive` section in the readme to clarify the minimum master nodes option.

yanns commented on incubator/elasticsearch/README.md:
```diff
@@ -108,6 +108,23 @@ would degrade performance heavily. The issue is tracked in
+
+# Client and Coordinating Nodes
+
+Elasticsearch v5 terminology has updated, and now refers to a `Client Node` as a `Coordianting Node`.
```
`Coordianting` -> `Coordinating`

icereval replied:
:+1:

yanns reviewed (changes requested):
Thx for the update!

mikesplain commented on incubator/elasticsearch/templates/client-svc.yaml:
```diff
@@ -12,6 +12,9 @@ spec:
   ports:
     - port: 9200
       targetPort: http
+  ports:
```
Looks like a copy and paste error, shouldn't have ports again.

mikesplain replied:
Also I think when you include multiple ports in a svc you need to add a name value to each of them.

icereval replied:
thanks for the catch, :+1:

mikesplain commented:
Great job though, I'm excited to get using this.

unguiculus commented on incubator/elasticsearch/values.yaml:
```diff
@@ -3,45 +3,52 @@
+  repository: "centerforopenscience/elasticsearch"
+  tag: "5.4"
+  # tag: "2.4"
   pullPolicy: "Always"
```
Default should be `IfNotPresent`.

unguiculus commented on incubator/elasticsearch/templates/master-statefulset.yaml:
```diff
@@ -50,6 +51,19 @@ spec:
           privileged: true
+      - name: "chown"
+        image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+        imagePullPolicy: {{ default "" .Values.image.pullPolicy | quote }}
```
No need to use `default ""` when default is set in `values.yaml`.

unguiculus commented:
I just figured that it would be even better to use the `appVersion` field in `Chart.yaml` file. The field is missing anyways. You can access it using `.Chart.AppVersion`.

unguiculus commented:
Marking this as stale. Please update within one week.

icereval commented:
@unguiculus, it was my impression the issues you brought up had been addressed [here](https://github.com/kubernetes/charts/pull/1794#discussion_r141850400).

unguiculus commented:
Please add the `appVersion` field to `Chart.yaml` instead of `values.yaml` and use it like so `{{ .Chart.appVersion }}`. That would be much better. We require this field for new charts anyways.

icereval commented:
My understanding is this field was immutable in the current build of helm and only used at package creation time, how would a user configure it for a 2.4 vs. 5.4 vs 5.5 deployment using the google master repo?

unguiculus commented:
Hm, fair enough, I didn't really think about that.

unguiculus commented:
Please fix conflicts, bump the chart version, and this should be good to go.

unguiculus commented on incubator/elasticsearch/README.md:
```diff
@@ -56,30 +56,35 @@ $ kubectl delete pvc -l release=my-release,component=data
-| `data.antiAffinity`                  | Data anti-affinity policy               | `soft`                              |
+|              Parameter               |                             Description                             |               Default                |
+| ------------------------------------ | ------------------------------------------------------------------- | ------------------------------------ |
+| `appVersion`                         | Application Version                                                 | `5.4`                                |
```
Please document what the field is used for, so people know when they may have to change it.

icereval replied:
added a deep dive on application version, https://github.com/kubernetes/charts/pull/1794/files#diff-4c83ed39abf993d7fded2101e2614aecR99

icereval committed:
Bump version and document appVersion

unguiculus commented:
/ok-to-test
